File: radar-commons/src/main/java/org/radarcns/producer/rest/AvroDataMapperFactory.java
package org.radarcns.producer.rest;

import static org.apache.avro.JsonProperties.NULL_VALUE;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.apache.avro.Schema;
import org.apache.avro.Schema.Type;
import org.apache.avro.SchemaValidationException;
import org.apache.avro.generic.GenericData;
import org.apache.avro.generic.GenericData.Fixed;
import org.apache.avro.generic.GenericEnumSymbol;
import org.apache.avro.generic.GenericRecord;
import org.apache.avro.generic.GenericRecordBuilder;
import org.apache.avro.generic.IndexedRecord;
import org.radarcns.util.Base64;
import org.radarcns.util.Base64.Encoder;

@SuppressWarnings({"PMD"})
public final class AvroDataMapperFactory {
    public static final AvroDataMapper IDENTITY_MAPPER = new AvroDataMapper() {
        @Override
        public Object convertAvro(Object obj) {
            return obj;
        }

        @Override
        public String toString() {
            return "Identity";
        }
    };
    private static final AvroDataMapperFactory INSTANCE = new AvroDataMapperFactory();

    public static AvroDataMapperFactory get() {
        return INSTANCE;
    }

    /**
     * Create a mapper for data in one Avro schema to that in another Avro schema.
     * @param from originating Avro schema
     * @param to resulting Avro schema
     * @param defaultVal default value as defined in an Avro record field,
     *                   may be null if there is no default value.
     * @return Avro data mapper
     * @throws SchemaValidationException if the given schemas are incompatible.
     */
    public AvroDataMapper createMapper(Schema from, Schema to, final Object defaultVal)
            throws SchemaValidationException {
        if (from.equals(to)) {
            return IDENTITY_MAPPER;
        }

        try {
            if (to.getType() == Schema.Type.UNION || from.getType() == Schema.Type.UNION) {
                return mapUnion(from, to, defaultVal);
            }
            if (to.getType() == Schema.Type.ENUM || to.getType() == Schema.Type.ENUM) {
                return mapEnum(from, to, defaultVal);
            }

            switch (to.getType()) {
                case INT:
                case LONG:
                case DOUBLE:
                case FLOAT:
                    return mapNumber(from, to, defaultVal);
                default:
                    break;
            }
            switch (from.getType()) {
                case RECORD:
                    return mapRecord(from, to);
                case ARRAY:
                    return mapArray(from, to);
                case MAP:
                    return mapMap(from, to);
                case FIXED:
                case BYTES:
                    return mapBytes(from, to, defaultVal);
                case INT:
                case LONG:
                case DOUBLE:
                case FLOAT:
                    return mapNumber(from, to, defaultVal);
                default:
                    if (from.getType() != to.getType()) {
                        throw new SchemaValidationException(to, from, new IllegalArgumentException(
                                "Schema types of from and to don't match"));
                    }
                    return IDENTITY_MAPPER;
            }
        } catch (SchemaValidationException ex) {
            if (defaultVal != null) {
                if (defaultVal == NULL_VALUE) {
                    return new AvroDataMapper() {
                        @Override
                        public Object convertAvro(Object obj) {
                            return null;
                        }
                    };
                } else {
                    return new AvroDataMapper() {
                        @Override
                        public Object convertAvro(Object obj) {
                            return defaultVal;
                        }
                    };
                }
            } else {
                throw ex;
            }
        }
    }

    /** Map one enum to another or to String. */
    private static AvroDataMapper mapEnum(Schema from, final Schema to, Object defaultVal)
            throws SchemaValidationException {
        if (to.getType() == Schema.Type.ENUM) {
            boolean containsAll = true;
            if (from.getType() == Schema.Type.ENUM) {
                for (String s : from.getEnumSymbols()) {
                    if (!to.hasEnumSymbol(s)) {
                        containsAll = false;
                        break;
                    }
                }
            } else if (from.getType() == Schema.Type.STRING) {
                containsAll = false;
            } else {
                throw new SchemaValidationException(to, from, new IllegalArgumentException(
                        "Cannot map enum from non-string or enum type"));
            }
            if (containsAll) {
                return new AvroDataMapper() {
                    @Override
                    public Object convertAvro(Object obj) {
                        return new GenericData.EnumSymbol(to, obj.toString());
                    }
                };
            } else {
                String defaultString = (String) defaultVal;
                if (defaultString == null && to.hasEnumSymbol("UNKNOWN")) {
                    defaultString = "UNKNOWN";
                }
                if (defaultString == null) {
                    throw new SchemaValidationException(to, from, new IllegalArgumentException(
                            "Cannot map enum symbols without default value"));
                } else {
                    final GenericEnumSymbol symbol = new GenericData.EnumSymbol(to, defaultString);
                    return new AvroDataMapper() {
                        @Override
                        public Object convertAvro(Object obj) {
                            String value = obj.toString();
                            if (to.hasEnumSymbol(value)) {
                                return new GenericData.EnumSymbol(to, value);
                            } else {
                                return symbol;
                            }
                        }
                    };
                }
            }
        } else if (from.getType() == Schema.Type.ENUM && to.getType() == Schema.Type.STRING) {
            return new AvroDataMapper() {
                @Override
                public Object convertAvro(Object obj) {
                    return obj.toString();
                }
            };
        } else {
            throw new SchemaValidationException(to, from, new IllegalArgumentException(
                    "Cannot map unknown type with enum."));
        }
    }

    /** Get the default value as a Generic type. */
    private static Object getDefaultValue(Object defaultVal, Schema schema) {
        if (defaultVal == null) {
            return null;
        } else if (schema.getType() == Schema.Type.ENUM) {
            return new GenericData.EnumSymbol(schema, defaultVal);
        } else {
            return defaultVal;
        }
    }

    /** Maps one number type to another or parses/converts to a string. */
    private static AvroDataMapper mapNumber(Schema from, Schema to, final Object defaultVal)
            throws SchemaValidationException {
        if (from.getType() == to.getType()) {
            return IDENTITY_MAPPER;
        }

        if (from.getType() == Schema.Type.STRING) {
            if (defaultVal == null) {
                throw new SchemaValidationException(to, from, new IllegalArgumentException(
                        "Cannot map string to number without default value."));
            } else {
                switch (to.getType()) {
                    case INT:
                        return new StringToNumberMapper(defaultVal) {
                            @Override
                            public Number stringToNumber(String obj) {
                                return Integer.valueOf(obj);
                            }
                        };
                    case LONG:
                        return new StringToNumberMapper(defaultVal) {
                            @Override
                            public Number stringToNumber(String obj) {
                                return Long.valueOf(obj);
                            }
                        };
                    case DOUBLE:
                        return new StringToNumberMapper(defaultVal) {
                            @Override
                            public Number stringToNumber(String obj) {
                                return Double.valueOf(obj);
                            }
                        };
                    case FLOAT:
                        return new StringToNumberMapper(defaultVal) {
                            @Override
                            public Number stringToNumber(String obj) {
                                return Float.valueOf(obj);
                            }
                        };
                    default:
                        throw new SchemaValidationException(to, from, new IllegalArgumentException(
                                "Cannot map numeric type with non-numeric type"));
                }
            }
        } else {
            switch (to.getType()) {
                case INT:
                    return new AvroDataMapper() {
                        @Override
                        public Object convertAvro(Object obj) {
                            return ((Number) obj).intValue();
                        }
                    };
                case LONG:
                    return new AvroDataMapper() {
                        @Override
                        public Object convertAvro(Object obj) {
                            return ((Number) obj).longValue();
                        }
                    };
                case DOUBLE:
                    return new AvroDataMapper() {
                        @Override
                        public Object convertAvro(Object obj) {
                            return Double.valueOf(obj.toString());
                        }
                    };
                case FLOAT:
                    return new AvroDataMapper() {
                        @Override
                        public Object convertAvro(Object obj) {
                            return ((Number) obj).floatValue();
                        }
                    };
                case STRING:
                    return new AvroDataMapper() {
                        @Override
                        public Object convertAvro(Object obj) {
                            return obj.toString();
                        }
                    };
                default:
                    throw new SchemaValidationException(to, from, new IllegalArgumentException(
                            "Cannot map numeric type with non-numeric type"));
            }
        }
    }

    /** Get the non-null union type of a nullable/optional union field. */
    private static Schema nonNullUnionSchema(Schema schema) throws SchemaValidationException {
        List<Schema> types = schema.getTypes();

        if (types.size() != 2) {
            throw new SchemaValidationException(schema, schema,
                    new IllegalArgumentException("Types must denote optionals"));
        }

        if (types.get(0).getType() == Schema.Type.NULL) {
            if (types.get(1).getType() != Schema.Type.NULL) {
                return types.get(1);
            } else {
                throw new SchemaValidationException(schema, schema,
                        new IllegalArgumentException("Types must denote optionals"));
            }
        } else if (types.get(1).getType() == Schema.Type.NULL) {
            return types.get(0);
        } else {
            throw new SchemaValidationException(schema, schema,
                    new IllegalArgumentException("Types must denote optionals."));
        }
    }

    /** Map one union to another, or a union to non-union, or non-union to union. */
    private AvroDataMapper mapUnion(Schema from, Schema to, Object defaultVal)
            throws SchemaValidationException {
        Schema resolvedFrom = from.getType() == Schema.Type.UNION ? nonNullUnionSchema(from) : from;

        if (from.getType() == Schema.Type.UNION && to.getType() != Schema.Type.UNION) {
            if (defaultVal != null) {
                final Object actualDefault = getDefaultValue(defaultVal, to);
                final AvroDataMapper subMapper = createMapper(resolvedFrom, to, defaultVal);
                return new AvroDataMapper() {
                    @Override
                    public Object convertAvro(Object obj) {
                        if (obj == null) {
                            return actualDefault;
                        } else {
                            return subMapper.convertAvro(obj);
                        }
                    }
                };
            } else {
                throw new SchemaValidationException(to, from, new IllegalArgumentException(
                        "Cannot map union to non-union without a default value"));
            }
        } else {
            Schema toNonNull = nonNullUnionSchema(to);
            final AvroDataMapper unionMapper = createMapper(resolvedFrom, toNonNull, defaultVal);
            return new AvroDataMapper() {
                @Override
                public Object convertAvro(Object obj) {
                    if (obj == null) {
                        return null;
                    } else {
                        return unionMapper.convertAvro(obj);
                    }
                }
            };
        }
    }

    /** Map an array to another. */
    private AvroDataMapper mapArray(Schema from, Schema to)
            throws SchemaValidationException {
        if (to.getType() != Schema.Type.ARRAY) {
            throw new SchemaValidationException(to, from,
                    new IllegalArgumentException("Cannot map array to non-array"));
        }
        final AvroDataMapper subMapper = createMapper(from.getElementType(), to.getElementType(),
                null);
        return new AvroDataMapper() {
            @Override
            public Object convertAvro(Object obj) {
                List array = (List) obj;
                List<Object> toArray = new ArrayList<>(array.size());
                for (Object val : array) {
                    toArray.add(subMapper.convertAvro(val));
                }
                return toArray;
            }
        };
    }

    /** Map a map to another. */
    private AvroDataMapper mapMap(Schema from, Schema to) throws SchemaValidationException {
        if (to.getType() != Schema.Type.MAP) {
            throw new SchemaValidationException(to, from,
                    new IllegalArgumentException("Cannot map array to non-array"));
        }
        final AvroDataMapper subMapper = createMapper(from.getValueType(), to.getValueType(),
                null);
        return new AvroDataMapper() {
            @Override
            public Object convertAvro(Object obj) {
                @SuppressWarnings("unchecked")
                Map<String, ?> map = (Map<String, ?>) obj;
                Map<String, Object> toMap = new HashMap<>(map.size() * 4 / 3 + 1);
                for (Map.Entry<String, ?> entry : map.entrySet()) {
                    toMap.put(entry.getKey(), subMapper.convertAvro(entry.getValue()));
                }
                return toMap;
            }
        };
    }

    private AvroDataMapper mapBytes(Schema from, final Schema to, final Object defaultVal)
            throws SchemaValidationException {
        if (from.getType() == to.getType()
                && (from.getType() == Type.BYTES
                || (from.getType() == Type.FIXED && from.getFixedSize() == to.getFixedSize()))) {
            return IDENTITY_MAPPER;
        } else if (from.getType() == Type.FIXED && to.getType() == Schema.Type.BYTES) {
            return new AvroDataMapper() {
                @Override
                public Object convertAvro(Object object) {
                    return ByteBuffer.wrap(((Fixed)object).bytes());
                }
            };
        } else if (from.getType() == Type.BYTES && to.getType() == Type.FIXED) {
            if (defaultVal == null) {
                throw new SchemaValidationException(to, from, new IllegalArgumentException(
                        "Cannot map bytes to fixed without default value"));
            }
            return new AvroDataMapper() {
                @Override
                public Object convertAvro(Object object) {
                    byte[] bytes = ((ByteBuffer) object).array();
                    if (bytes.length == to.getFixedSize()) {
                        return GenericData.get().createFixed(null, bytes, to);
                    } else {
                        return GenericData.get().createFixed(null, (byte[]) defaultVal, to);
                    }
                }
            };
        } else if (to.getType() == Type.STRING) {
            final Encoder encoder = Base64.getEncoder();
            if (from.getType() == Type.FIXED) {
                return new AvroDataMapper() {
                    @Override
                    public Object convertAvro(Object object) {
                        return new String(encoder.encode(((Fixed) object).bytes()),
                                StandardCharsets.UTF_8);
                    }
                };
            } else {
                return new AvroDataMapper() {
                    @Override
                    public Object convertAvro(Object object) {
                        return new String(encoder.encode(((ByteBuffer) object).array()),
                                StandardCharsets.UTF_8);
                    }
                };
            }
        } else {
            throw new SchemaValidationException(to, from,
                    new IllegalArgumentException(
                            "Fixed type must be mapped to comparable byte size"));
        }
    }


    private AvroDataMapper mapRecord(Schema from, Schema to)
        throws SchemaValidationException {
        if (to.getType() != Schema.Type.RECORD) {
            throw new SchemaValidationException(to, from,
                new IllegalArgumentException("From and to schemas must be records."));
        }
        List<Schema.Field> fromFields = from.getFields();
        Schema.Field[] toFields = new Schema.Field[fromFields.size()];
        AvroDataMapper[] fieldMappers = new AvroDataMapper[fromFields.size()];

        boolean[] filledPositions = new boolean[to.getFields().size()];

        for (int i = 0; i < fromFields.size(); i++) {
            Schema.Field fromField = fromFields.get(i);
            Schema.Field toField = to.getField(fromField.name());
            if (toField == null) {
                continue;
            }

            filledPositions[toField.pos()] = true;

            Schema fromSchema = fromField.schema();
            Schema toSchema = toField.schema();

            toFields[i] = toField;
            fieldMappers[i] = createMapper(fromSchema, toSchema, toField.defaultVal());
        }

        for (int i = 0; i < filledPositions.length; i++) {
            if (!filledPositions[i] && to.getFields().get(i).defaultVal() == null) {
                throw new SchemaValidationException(to, from,
                    new IllegalArgumentException("Cannot map to record without default value"
                            + " for new field " + to.getFields().get(i).name()));
            }
        }

        return new RecordMapper(to, toFields, fieldMappers);
    }

    /** Maps one record to another. */
    private static class RecordMapper implements AvroDataMapper {
        private final AvroDataMapper[] fieldMappers;
        private final Schema.Field[] toFields;
        private final Schema toSchema;

        RecordMapper(Schema toSchema, Schema.Field[] toFields, AvroDataMapper[] fieldMappers) {
            this.toSchema = toSchema;
            this.fieldMappers = fieldMappers;
            this.toFields = toFields;
        }


        @Override
        public GenericRecord convertAvro(Object obj) {
            GenericRecordBuilder builder = new GenericRecordBuilder(toSchema);
            IndexedRecord record = (IndexedRecord) obj;
            for (int i = 0; i < toFields.length; i++) {
                Schema.Field field = toFields[i];
                if (field == null) {
                    continue;
                }
                builder.set(field, fieldMappers[i].convertAvro(record.get(i)));
            }
            return builder.build();
        }

        @Override
        public String toString() {
            return "RecordMapper{"
                    + "fieldMappers=" + Arrays.toString(fieldMappers)
                    + ", toFields=" + Arrays.toString(toFields) + '}';
        }
    }

    private abstract static class StringToNumberMapper implements AvroDataMapper {
        private final Object defaultVal;

        StringToNumberMapper(Object defaultVal) {
            this.defaultVal = defaultVal;
        }

        @Override
        public Object convertAvro(Object object) {
            try {
                return stringToNumber(object.toString());
            } catch (NumberFormatException ex) {
                return defaultVal;
            }
        }

        abstract Number stringToNumber(String toString);
    }
}

File: radar-commons/src/main/java/org/radarbase/producer/avro/AvroDataMapperFactory.kt
package org.radarbase.producer.avro

import org.apache.avro.JsonProperties
import org.apache.avro.Schema
import org.apache.avro.SchemaValidationException
import org.apache.avro.generic.*
import org.radarbase.util.Base64Encoder
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.nio.ByteBuffer
import java.util.*

object AvroDataMapperFactory {
    /**
     * Create a mapper for data in one Avro schema to that in another Avro schema.
     * @param from originating Avro schema
     * @param to resulting Avro schema
     * @param defaultVal default value as defined in an Avro record field,
     * may be null if there is no default value.
     * @return Avro data mapper
     * @throws SchemaValidationException if the given schemas are incompatible.
     */
    @Throws(SchemaValidationException::class)
    fun createMapper(from: Schema, to: Schema, defaultVal: Any?): AvroDataMapper {
        if (from == to) {
            logger.debug("Using identity schema mapping from {} to {}", from, to)
            return IDENTITY_MAPPER
        }
        logger.debug("Computing custom mapping from {} to {}", from, to)
        return try {
            if (to.type == Schema.Type.UNION || from.type == Schema.Type.UNION) {
                return mapUnion(from, to, defaultVal)
            }
            if (to.type == Schema.Type.ENUM || from.type == Schema.Type.ENUM) {
                return mapEnum(from, to, defaultVal)
            }
            when (to.type) {
                Schema.Type.INT, Schema.Type.LONG, Schema.Type.DOUBLE, Schema.Type.FLOAT ->
                    return mapNumber(from, to, defaultVal)
                else -> {}
            }
            when (from.type) {
                Schema.Type.RECORD -> mapRecord(from, to)
                Schema.Type.ARRAY -> mapArray(from, to)
                Schema.Type.MAP -> mapMap(from, to)
                Schema.Type.FIXED, Schema.Type.BYTES -> mapBytes(from, to, defaultVal)
                Schema.Type.INT, Schema.Type.LONG, Schema.Type.DOUBLE, Schema.Type.FLOAT ->
                    mapNumber(from, to, defaultVal)
                to.type -> IDENTITY_MAPPER
                else -> throw validationException(to, from, "Schema types of from and to don't match")
            }
        } catch (ex: SchemaValidationException) {
            defaultVal ?: throw ex
            if (defaultVal === JsonProperties.NULL_VALUE) {
                AvroDataMapper { null }
            } else {
                AvroDataMapper { defaultVal }
            }
        }
    }

    /** Map one union to another, or a union to non-union, or non-union to union.  */
    @Throws(SchemaValidationException::class)
    private fun mapUnion(from: Schema, to: Schema, defaultVal: Any?): AvroDataMapper {
        // Do not create a custom mapper for trivial changes.
        if (
            from.type == Schema.Type.UNION &&
            to.type == Schema.Type.UNION &&
            from.types.size == from.types.size
        ) {
            val matches = from.types.indices.all { i ->
                val fromType = from.types[i].type
                val toType = to.types[i].type
                fromType == toType && fromType.isPrimitive()
            }
            if (matches) {
                return IDENTITY_MAPPER
            }
        }
        val resolvedFrom = if (from.type == Schema.Type.UNION) {
            nonNullUnionSchema(from)
        } else {
            from
        }

        return if (from.type == Schema.Type.UNION && to.type != Schema.Type.UNION) {
            defaultVal ?: throw validationException(to, from, "Cannot map union to non-union without a default value")
            val actualDefault = getDefaultValue(defaultVal, to)
            val subMapper = createMapper(resolvedFrom, to, defaultVal)
            AvroDataMapper { obj ->
                if (obj == null) {
                    actualDefault
                } else {
                    subMapper.convertAvro(obj)
                }
            }
        } else {
            val toNonNull = nonNullUnionSchema(to)
            val unionMapper = createMapper(resolvedFrom, toNonNull, defaultVal)
            AvroDataMapper { obj ->
                obj ?: return@AvroDataMapper null
                unionMapper.convertAvro(obj)
            }
        }
    }

    /** Map an array to another.  */
    @Throws(SchemaValidationException::class)
    private fun mapArray(from: Schema, to: Schema): AvroDataMapper {
        if (to.type != Schema.Type.ARRAY) {
            throw validationException(to, from, "Cannot map array to non-array")
        }
        val subMapper = createMapper(from.elementType, to.elementType, null)
        return AvroDataMapper { obj ->
            obj.asAvroType<List<*>>(from, to).map { subMapper.convertAvro(it) }
        }
    }

    /** Map a map to another.  */
    @Throws(SchemaValidationException::class)
    private fun mapMap(from: Schema, to: Schema): AvroDataMapper {
        if (to.type != Schema.Type.MAP) {
            throw validationException(to, from, "Cannot map map to non-map")
        }
        val subMapper = createMapper(from.valueType, to.valueType, null)
        return AvroDataMapper { obj ->
            buildMap {
                obj.asAvroType<Map<*, *>>(from, to).forEach { (k, v) ->
                    put(k.toString(), subMapper.convertAvro(v))
                }
            }
        }
    }

    @Throws(SchemaValidationException::class)
    private fun mapBytes(from: Schema, to: Schema, defaultVal: Any?): AvroDataMapper {
        return if (from.type == Schema.Type.BYTES && to.type == Schema.Type.BYTES) {
            IDENTITY_MAPPER
        } else if (from.type == Schema.Type.FIXED && to.type == Schema.Type.FIXED &&
            from.fixedSize == to.fixedSize
        ) {
            IDENTITY_MAPPER
        } else if (from.type == Schema.Type.FIXED && to.type == Schema.Type.BYTES) {
            AvroDataMapper { `object` ->
                ByteBuffer.wrap(`object`.asAvroType<GenericData.Fixed>(from, to).bytes())
            }
        } else if (from.type == Schema.Type.BYTES && to.type == Schema.Type.FIXED) {
            defaultVal ?: throw validationException(to, from, "Cannot map bytes to fixed without default value")

            AvroDataMapper { `object`: Any? ->
                val bytes = `object`.asAvroType<ByteBuffer>(from, to).array()
                val value = if (bytes.size == to.fixedSize) {
                    bytes
                } else {
                    defaultVal as? ByteArray
                }
                GenericData.get().createFixed(null, value, to)
            }
        } else if (to.type == Schema.Type.STRING) {
            val encoder = Base64Encoder
            if (from.type == Schema.Type.FIXED) {
                AvroDataMapper { `object` ->
                    encoder.encode(`object`.asAvroType<GenericData.Fixed>(from, to).bytes())
                }
            } else {
                AvroDataMapper { `object` ->
                    encoder.encode(`object`.asAvroType<ByteBuffer>(from, to).array())
                }
            }
        } else {
            throw validationException(to, from, "Fixed type must be mapped to comparable byte size")
        }
    }

    @Throws(SchemaValidationException::class)
    private fun mapRecord(from: Schema, to: Schema): AvroDataMapper {
        if (to.type != Schema.Type.RECORD) {
            throw validationException(to, from, "From and to schemas must be records.")
        }
        val fromFields = from.fields
        val toFields = arrayOfNulls<Schema.Field?>(
            fromFields.size,
        )
        val fieldMappers = arrayOfNulls<AvroDataMapper?>(
            fromFields.size,
        )
        val filledPositions = BooleanArray(to.fields.size)
        for (i in fromFields.indices) {
            val fromField = fromFields[i]
            val toField = to.getField(fromField.name()) ?: continue
            filledPositions[toField.pos()] = true
            toFields[i] = toField
            fieldMappers[i] = createMapper(
                fromField.schema(),
                toField.schema(),
                toField.defaultVal(),
            )
        }
        filledPositions.forEachIndexed { i, isFilled ->
            if (!isFilled && to.fields[i].defaultVal() == null) {
                throw validationException(
                    to,
                    from,
                    "Cannot map to record without default value for new field ${to.fields[i].name()}",
                )
            }
        }
        return RecordMapper(to, toFields, fieldMappers)
    }

    /** Maps one record to another.  */
    private class RecordMapper constructor(
        private val toSchema: Schema,
        private val toFields: Array<Schema.Field?>,
        private val fieldMappers: Array<AvroDataMapper?>,
    ) : AvroDataMapper {
        override fun convertAvro(`object`: Any?): GenericRecord {
            val builder = GenericRecordBuilder(toSchema)
            val record = `object`.asAvroType<IndexedRecord>(toSchema, toSchema)
            for (i in toFields.indices) {
                val field = toFields[i] ?: continue
                val mapper = fieldMappers[i] ?: continue
                builder[field] = mapper.convertAvro(record[i])
            }
            return builder.build()
        }

        override fun toString(): String {
            return (
                "RecordMapper{" +
                    "fieldMappers=" + fieldMappers.contentToString() +
                    ", toFields=" + toFields.contentToString() + '}'
                )
        }
    }

    private class StringToNumberMapper(
        private val defaultVal: Any?,
        private val mapping: (String) -> Number,
    ) :
        AvroDataMapper {
        override fun convertAvro(`object`: Any?): Any? {
            `object` ?: return defaultVal
            return try {
                mapping(`object`.toString())
            } catch (ex: NumberFormatException) {
                defaultVal
            }
        }
    }

    private val logger: Logger = LoggerFactory.getLogger(AvroDataMapperFactory::class.java)
    val IDENTITY_MAPPER: AvroDataMapper = object : AvroDataMapper {
        override fun convertAvro(`object`: Any?): Any? = `object`

        override fun toString(): String = "Identity"
    }

    private inline fun <reified T> Any?.asAvroType(from: Schema, to: Schema): T {
        if (this !is T) {
            throw validationException(
                to,
                from,
                "${to.type} type cannot be mapped from ${this?.javaClass?.name} Java type.",
            )
        }
        return this
    }

    private val PRIMITIVE_TYPES = EnumSet.of(
        Schema.Type.INT,
        Schema.Type.LONG,
        Schema.Type.BYTES,
        Schema.Type.FLOAT,
        Schema.Type.DOUBLE,
        Schema.Type.NULL,
        Schema.Type.BOOLEAN,
        Schema.Type.STRING,
    )

    /** Map one enum to another or to String.  */
    @Throws(SchemaValidationException::class)
    private fun mapEnum(from: Schema, to: Schema, defaultVal: Any?): AvroDataMapper {
        return if (to.type == Schema.Type.ENUM) {
            var containsAll = true
            if (from.type == Schema.Type.ENUM) {
                for (s in from.enumSymbols) {
                    if (!to.hasEnumSymbol(s)) {
                        containsAll = false
                        break
                    }
                }
            } else if (from.type == Schema.Type.STRING) {
                containsAll = false
            } else {
                throw validationException(to, from, "Cannot map enum from non-string or enum type")
            }
            if (containsAll) {
                AvroDataMapper { obj -> GenericData.EnumSymbol(to, obj.toString()) }
            } else {
                var defaultString = defaultVal as? String
                if (defaultString == null) {
                    if (to.hasEnumSymbol("UNKNOWN")) {
                        defaultString = "UNKNOWN"
                    } else {
                        throw validationException(
                            to,
                            from,
                            "Cannot map enum symbols without default value",
                        )
                    }
                }
                val symbol: GenericEnumSymbol<*> = GenericData.EnumSymbol(to, defaultString)
                AvroDataMapper { obj: Any? ->
                    val value = obj.toString()
                    if (to.hasEnumSymbol(value)) {
                        GenericData.EnumSymbol(to, value)
                    } else {
                        symbol
                    }
                }
            }
        } else if (from.type == Schema.Type.ENUM && to.type == Schema.Type.STRING) {
            AvroDataMapper { it.toString() }
        } else {
            throw validationException(to, from, "Cannot map unknown type with enum.")
        }
    }

    /** Get the default value as a Generic type.  */
    private fun getDefaultValue(defaultVal: Any?, schema: Schema): Any? {
        return if (defaultVal == null) {
            null
        } else if (schema.type == Schema.Type.ENUM) {
            GenericData.EnumSymbol(schema, defaultVal)
        } else {
            defaultVal
        }
    }

    /** Maps one number type to another or parses/converts to a string.  */
    @Throws(SchemaValidationException::class)
    private fun mapNumber(from: Schema, to: Schema, defaultVal: Any?): AvroDataMapper {
        if (from.type == to.type) {
            return IDENTITY_MAPPER
        }
        return if (from.type == Schema.Type.STRING) {
            defaultVal ?: throw validationException(to, from, "Cannot map string to number without default value.")
            when (to.type) {
                Schema.Type.INT -> StringToNumberMapper(defaultVal, Integer::valueOf)
                Schema.Type.LONG -> StringToNumberMapper(defaultVal, String::toLong)
                Schema.Type.DOUBLE -> StringToNumberMapper(defaultVal, String::toDouble)
                Schema.Type.FLOAT -> StringToNumberMapper(defaultVal, String::toFloat)
                else -> throw validationException(
                    to,
                    from,
                    "Cannot map numeric type with non-numeric type",
                )
            }
        } else {
            when (to.type) {
                Schema.Type.INT -> AvroDataMapper { it.asAvroType<Number>(from, to).toInt() }
                Schema.Type.LONG -> AvroDataMapper { it.asAvroType<Number>(from, to).toLong() }
                Schema.Type.DOUBLE -> AvroDataMapper { it.toString().toDouble() }
                Schema.Type.FLOAT -> AvroDataMapper { it.asAvroType<Number>(from, to).toFloat() }
                Schema.Type.STRING -> AvroDataMapper { it.toString() }
                else -> throw validationException(
                    to,
                    from,
                    "Cannot map numeric type with non-numeric type",
                )
            }
        }
    }

    /** Get the non-null union type of a nullable/optional union field.  */
    @Throws(SchemaValidationException::class)
    private fun nonNullUnionSchema(schema: Schema): Schema {
        val types = checkNotNull(schema.types) { "Union does not have subtypes" }
        if (types.size != 2) {
            throw validationException(schema, schema, "Types must denote optionals.")
        }
        return if (types[0].type == Schema.Type.NULL) {
            if (types[1].type != Schema.Type.NULL) {
                types[1]
            } else {
                throw validationException(schema, schema, "Types must denote optionals.")
            }
        } else if (types[1].type == Schema.Type.NULL) {
            types[0]
        } else {
            throw validationException(schema, schema, "Types must denote optionals.")
        }
    }

    private fun Schema.Type.isPrimitive(): Boolean = this in PRIMITIVE_TYPES

    internal fun validationException(
        from: Schema,
        to: Schema,
        message: String,
    ): SchemaValidationException = SchemaValidationException(
        to,
        from,
        IllegalArgumentException(message),
    )
}
